### PR TITLE
Add partial support for indexed TGA files

### PR DESCRIPTION
**Pull request by tamas-rabel**
Add support for TGA files with 24-bit palettes.

### CODE REVIEW

**Comment by tamas-rabel** [comment]
Extra context: Although the indexed TGA format seems archaic, Substance Painter exports all grayscale TGAs as indexed.

**Comment by walbourn** [comment]
Thanks for the contribution! Can you attach some example TGA files so I can include them in the test suite?

**Comment by tamas-rabel** [comment]
Sure, I've created indexed version of two of the test files from the test repo.
Thanks.

[indexed_tga_files.zip](https://github.com/microsoft/DirectXTex/files/8610614/indexed_tga_files.zip)

**Comment by walbourn** [comment]
Thanks for the test images. I'll have to sit on the PR a bit... I need to see how I go about getting the updated reader fuzz tested...

**Comment by tamas-rabel** [comment]
Ok, thanks.

File: DirectXTex/DirectXTexTGA.cpp [hunk 1]
```diff
@@ -258,6 +264,59 @@ namespace
...
+        {
+            return HRESULT_E_INVALID_DATA;
+        }
+
+        const uint8_t* bytes = reinterpret_cast<const uint8_t*>(pSource);
+
+        for (int i = 0; i < pHeader->wColorMapLength; i++)
+        {
+            switch (pHeader->bColorMapSize)
+            {
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:299]: For security purposes, have to validate bColorMapSize *before* using it in a computation. I suggest doing it up in DecodeHeader.

File: DirectXTex/DirectXTexTGA.cpp [hunk 2]
```diff
@@ -258,6 +264,59 @@ namespace
     }
 
 
+    //-------------------------------------------------------------------------------------
+    // Reads palette for paletted TGA formats
+    //-------------------------------------------------------------------------------------
+    HRESULT ReadPalette(
+        uint8_t* header,
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:271]: const and SAL

File: DirectXTex/DirectXTexTGA.cpp [hunk 3]
```diff
@@ -1953,13 +2045,28 @@ HRESULT DirectX::LoadFromTGAFile(
...
+        size_t palette_offset = 0;
+        if (convFlags & CONV_FLAGS_PALETTED)
+        {
+            auto pHeader = reinterpret_cast<const TGA_HEADER*>(header);
+            palette_offset += pHeader->wColorMapLength * (pHeader->bColorMapSize / 8);
+        }
+
         if (convFlags & CONV_FLAGS_RLE)
         {
-            hr = UncompressPixels(temp.get(), remaining, flags, image.GetImage(0, 0, 0), convFlags);
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:1958]: remaining also needs adjustment: ``remaining - palette_offset``

File: DirectXTex/DirectXTexTGA.cpp [hunk 4]
```diff
@@ -258,6 +264,59 @@ namespace
...
+            return HRESULT_E_NOT_SUPPORTED;
+        }
+        if (size_t(pHeader->wColorMapLength * (pHeader->bColorMapSize / 8)) > size)
+        {
+            return HRESULT_E_INVALID_DATA;
+        }
+
+        const uint8_t* bytes = reinterpret_cast<const uint8_t*>(pSource);
+
+        for (int i = 0; i < pHeader->wColorMapLength; i++)
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:296]: Assuming ``wColorMapFirst`` is 0 here.

File: DirectXTex/DirectXTexTGA.cpp [hunk 5]
```diff
@@ -258,6 +264,59 @@ namespace
...
+        if (pHeader->wColorMapLength > 256)
+        {
+            return HRESULT_E_NOT_SUPPORTED;
+        }
+        if (size_t(pHeader->wColorMapLength * (pHeader->bColorMapSize / 8)) > size)
+        {
+            return HRESULT_E_INVALID_DATA;
+        }
+
+        const uint8_t* bytes = reinterpret_cast<const uint8_t*>(pSource);
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:294]: ``auto`` is best in situations like there where you are just replicating the type on LHS and RHS.

**Review by walbourn — CHANGES_REQUESTED** [review]
I'll do these changes when I pull in the PR :)

File: DirectXTex/DirectXTexTGA.cpp [hunk 6]
```diff
@@ -225,9 +220,20 @@ namespace
             break;
 
         case TGA_NO_IMAGE:
+            return HRESULT_E_NOT_SUPPORTED;
         case TGA_COLOR_MAPPED:
         case TGA_COLOR_MAPPED_RLE:
-            return HRESULT_E_NOT_SUPPORTED;
+            metadata.format = DXGI_FORMAT_R8G8B8A8_UNORM_SRGB;
```
[the remark this answers is not in the document]
Comment (walbourn) [reply DirectXTex/DirectXTexTGA.cpp:226]: Also should check ``TGA_FLAGS_BGR``.

File: DirectXTex/DirectXTexTGA.cpp [hunk 7]
```diff
@@ -258,6 +264,59 @@ namespace
...
+        const uint8_t* bytes = reinterpret_cast<const uint8_t*>(pSource);
+
+        for (int i = 0; i < pHeader->wColorMapLength; i++)
+        {
+            switch (pHeader->bColorMapSize)
+            {
+            case 16:
+                return HRESULT_E_NOT_SUPPORTED;
+            case 24:
+                palette[i * 4 + 0] = bytes[2];
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:303]: The spec seems to indicate colors map data is RGB, but you are reading it wrong as BGR.  Need to try out a color image. ``ucm8.tga`` in my directxtexmedia should work. It's yellow according to DxTex.

Comment (walbourn) [reply DirectXTex/DirectXTexTGA.cpp:303]: D3DX code says BGR for colormaps...

File: DirectXTex/DirectXTexTGA.cpp [hunk 8]
```diff
@@ -814,7 +873,8 @@ namespace
         size_t size,
         TGA_FLAGS flags,
         _In_ const Image* image,
-        _In_ uint32_t convFlags) noexcept
+        _In_ uint32_t convFlags,
+        _In_ uint8_t* palette) noexcept
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:877]: Since this can be nullptr, ``_In_opt_``

File: DirectXTex/DirectXTexTGA.cpp [hunk 9]
```diff
@@ -1698,7 +1790,7 @@ HRESULT DirectX::LoadFromTGAFile(
```
Comment (walbourn) [inline DirectXTex/DirectXTexTGA.cpp:1790]: Also need to implement the same thing for TGA load from memory.

**Comment by walbourn** [comment]
Cleaned up and committed to main https://github.com/microsoft/DirectXTex/commit/99518756031394c8cc4b3d5c308b2795dc8e8e9a

**Pull request closed without merging**